torch/tensortypes/nestedtensor/nested.py

```diff
@@ -0,0 +1,215 @@
+                        self.dtype == tensor.dtype and
+                        requires_grad == tensor.requires_grad and
+                        is_pinned == tensor.is_pinned()):
+                    raise ValueError("Each passed Tensor "
+                                     "must match in dim, layout, "
+                                     "device, dtype and requires_grad")
+        else:
+            # Carrying around information as member variables vs.
+            # checking one entry of the owned Tensors is annoying
+            # and error-prone. Carrying around an is_empty attribute
+            # to hide the fact that we carry around a list with a
+            # single empty Tensor is also annoying and error-prone.
+            # Both are not worth it for a minor feature.
+            raise ValueError("We do not support empty lists for now.")
+
+    def __getattribute__(self, attr):
+        if attr == 'shape':
+            raise NotImplementedError()
+        if attr == 'requires_grad':
+            return self.tensors[0].requires_grad
+        if attr == 'grad_fn':
+            # We're assuming len(self.tensors) is non-zero
+            if self.tensors[0].grad_fn is None:
+                for i in range(1, len(self.tensors)):
+                    assert self.tensors[i].grad_fn is None
+                return None
+
+            def create_grad_fn(nested_tensor):
+                def _func(*args, **kwargs):
+                    for tensor in nested_tensor.tensors:
+                        tensor.grad_fn(*args, **kwargs)
+                return _func
+            return create_grad_fn(self)
+        if attr == 'grad':
+            grads = []
+            for tensor in self.tensors:
+                grads.append(tensor.grad)
+            if all(grad is None for grad in grads):
+                return None
+            else:
+                return NestedTensor(grads)
+        if attr == 'data':
+            data = []
+            for tensor in self.tensors:
+                data.append(tensor.data)
+            return NestedTensor(data)
+        return super().__getattribute__(attr)
+
+    def __len__(self):
+        return len(self.tensors)
+
+    def __bool__(self):
+        raise NotImplementedError(
+            "This has not been covered by NestedTensor 0.0.1")
+
+    def __str__(self):
+        tensors = self.unbind()
+        result = "nestedtensor([\n"
+        for tensor in tensors:
+            result += "  " + tensor.__str__() + "\n"
+        result += "])"
+        return result
+
+    def __repr__(self):
+        tensors = self.unbind()
+        result = "nestedtensor([\n"
+        for tensor in tensors:
+            result += "  " + tensor.__repr__() + "\n"
+        result += "])"
+        return result
+
+    def is_empty(self):
+        # This condition can never be true, since we disallow an empty list for now.
+        raise ValueError("self.tensors cannot be empty under current constraints.")
```

Comment:
Shouldn't you just return `False`? Or maybe this should be tensorwise?

Reply:
If the user expects it to be empty, but it must never be empty upon construction, then this is the kind of noisy failure I would like to see. Because this means something must have gone very wrong in either the user assumptions or program.